### PR TITLE
New user register flow

### LINKED ISSUE #1091: Novo fluxo de cadastro de usuários
## Feature Description

- Atualização de layout
- Após validação será solicitado para preencher os dados pessoais principais (Nome, telefone, cidade e estado)
- Cadastro via integração OAUTH com Google e Facebook será incluso em uma issue a parte.

Cadastro - Passo 1 - Forma de acesso

Cadastro - Passo 2 - Credenciais

Cadastro - Passo 3 - Validação

Cadastro - Passo 4 - Dados Pessoais

Cadastro - Passo 5 - Finalização

Mobile

MOB - Cadastro de Usuário - Passo 1

MOB - Cadastro de Usuário - Passo 2

MOB - Cadastro de Usuário - Passo 2 - ERROR

MOB - Cadastro de Usuário - Passo 3

MOB - Cadastro de Usuário - Passo 4

MOB - Cadastro de Usuário - Confirmação

---------------

## Acceptance criteria

* 

## Implementation Brief

*

### PR DESCRIPTION
**Pull request by alexandreadames**
## Summary

Resolves #1091 

### New User register flow:

1. Register route:

`POST /auth/register` 

Body:

```json
{
   "email": "newuser_reg_flow@gmail.com",
   "password": "123456",
   "disclaimers":[1,2,3,4,5,6,7]
}
```

User receive an email with token code

2. Confirm Account

`POST /auth/confirm-account`

Body:

```json
{
   "email":"newuser_reg_flow@gmail.com",
   "token":"260428"
}
```

This route returns a JWT token, this token should be used in next step.

3. User updates profile data

`POST users/:id`

Body:

```json
{
   "full_name": "New User registration flow",
   "phone": "(84) 98823-7668",
   "city": "Mossoró",
   "state": "RN"
}
```

### City_id and state_id migrated to user schema

Users with cities => 48
Migrated user cities => 46 (95.83%)
Migrated user states => 45 (93.75%)

## QA Steps

Run `adonis test --files test/functional/auth.spec.js`, test: `/auth/register user register flow`

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code passes the linting.
- [x] My code has proper inline documentation.
- [x] I have manually tested this PR.

### CODE REVIEW

File: packages/api/app/Controllers/Http/AuthController.js [hunk 1]
```diff
@@ -67,15 +66,15 @@ class AuthController {
 		user.status = 'verified';
 		await user.save();
 
-		const mailData = {
+		/* const mailData = {
```
Comment (alexandreadames) [inline packages/api/app/Controllers/Http/AuthController.js:69]: Is it necessary this confirmation mail now?

Comment (mateus4k) [reply packages/api/app/Controllers/Http/AuthController.js:69]: I think it is not necessary

**Review by mateus4k — APPROVED** [review]
Everything looks good 👍

File: packages/api/database/migrations/1626211781279_user_city_state_schema.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+				.onUpdate('CASCADE')
+				.onDelete('SET NULL');
+		});
+		this.schedule(async (trx) => {
+			const usersWithCity = await Database.table('users')
+				.whereNotNull('city')
+				.transacting(trx);
+
+			for await (const user of usersWithCity) {
+				// Tryng update user city
```
Comment (mateus4k) [inline packages/api/database/migrations/1626211781279_user_city_state_schema.js:32]:
```suggestion
				// Trying to update the user's city
```

File: packages/api/database/migrations/1626211781279_user_city_state_schema.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+					.whereRaw(`UPPER(name) = UPPER('${user.city}')`)
+					.first();
+
+				if (userCity) {
+					await Database.table('users')
+						.where('id', user.id)
+						.update({ city_id: userCity.id });
+				}
+
+				// Tryng update user state
```
Comment (mateus4k) [inline packages/api/database/migrations/1626211781279_user_city_state_schema.js:43]:
```suggestion
				// Trying to update the user's state
```

File: packages/api/database/migrations/1626211781279_user_city_state_schema.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+				const userState = await Database.table('states')
+					.whereRaw(
+						`UPPER(name) = UPPER('${user.state}') OR UPPER(initials) = UPPER('${user.state}')`,
+					)
```
Comment (mateus4k) [inline packages/api/database/migrations/1626211781279_user_city_state_schema.js:47]:
```suggestion
					.whereRaw('UPPER(name) = UPPER(?) OR UPPER(initials) = UPPER(?)', [user.state, user.state])
```

File: packages/api/database/factory.js [hunk 5]
```diff
@@ -49,8 +49,6 @@ Factory.blueprint('App/Models/User', async (faker, i, data) => {
 		address: faker.string(),
 		address2: faker.string(),
 		district: faker.string(),
-		city: faker.string(),
-		state: faker.string(),
```
Comment (mateus4k) [inline packages/api/database/factory.js:53]: You should add new fields here instead of removing old ones

Comment (alexandreadames) [reply packages/api/database/factory.js:53]: In factory, we often don't include fields with foreign key. If I desire create a new user with a city i can make:
`const userFactory = await Factory.model('App/Models/user').create({city_id:1111});`

File: packages/api/test/functional/announcement.spec.js [hunk 6]
```diff
@@ -139,6 +139,8 @@ test('POST /announcements creates a new Announcement', async ({ client, assert }
 		})
 		.end();
 
+	console.log(response.error);
```
Comment (mateus4k) [inline packages/api/test/functional/announcement.spec.js:142]: 👁️

**Review by mateus4k — DISMISSED** [review]
Approved! ✔️ 

Please mark as "resolved" review comments that you have already resolved. This helps reviewers a lot to know what problems still have a pending solution.

image

**Comment by mateus4k** [comment]
Have you already tested running cities seed (`npm run seed:cities`) after already have users depending on city IDs in your database?

It's necessary to avoid errors if in the future we need to run this seed again 😬

**Comment by alexandreadames** [comment]
> Have you already tested running cities seed (`npm run seed:cities`) after already have users depending on city IDs in your database?
> 
> It's necessary to avoid errors if in the future we need to run this seed again 😬

After running CitySeeder the city_id and state_id in user schema was deleted: `onDelete('SET NULL')`. This is not good because we going to lost info.

**Review by mateus4k — APPROVED** [review]

**Pull request merged**